### PR TITLE
Integrate Brave omaha

### PR DESCRIPTION
**Pull request by simonhong**
Testing with https://github.com/simonhong/omaha/tree/brave_omaha

When refresh `settings/help`, BraveUpdate.exe process runs and terminated.

Work together with https://github.com/brave/brave-browser/pull/303.

Closes: https://github.com/brave/brave-browser/issues/179

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions

### CODE REVIEW

File: patches/chrome-installer-util-util_constants.cc.patch [hunk 1]
```diff
@@ -1,8 +1,22 @@
...
 +++ b/chrome/installer/util/util_constants.cc
-@@ -187,7 +187,11 @@ const char kGoogleUpdateIsMachineEnvVar[] = "GoogleUpdateIsMachine";
+@@ -171,7 +171,13 @@ namespace env_vars {
+ // The presence of this environment variable with a value of 1 implies that
+ // setup.exe should run as a system installation regardless of what is on the
+ // command line.
++#if defined(BRAVE_CHROMIUM_BUILD) && defined(OFFICIAL_BUILD)
++// Reflect brave branded-omaha variable name.
++// Company name(BraveSoftware) + Product name(Update).
++const char kGoogleUpdateIsMachineEnvVar[] = "BraveSoftwareUpdateIsMachine";
```
Comment (bridiver) [inline patches/chrome-installer-util-util_constants.cc.patch]: is this necessary? I think it would be better to have consistent behavior for this env var with Chrome

File: patches/components-update_client-updater_state_win.cc.patch [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
++++ b/components/update_client/updater_state_win.cc
+@@ -23,6 +23,22 @@ namespace update_client {
+ 
+ namespace {
+ 
++#if defined(BRAVE_CHROMIUM_BUILD)
++const wchar_t kGoogleUpdatePoliciesKey[] =
++    L"SOFTWARE\\Policies\\BraveSoftware\\Update";
++const wchar_t kCheckPeriodOverrideMinutes[] = L"AutoUpdateCheckPeriodMinutes";
++const wchar_t kUpdatePolicyValue[] = L"UpdateDefault";
```
Comment (bridiver) [inline patches/components-update_client-updater_state_win.cc.patch]: most of these values look identical to the google versions. Seems like it would be better to use the undefine/redefine override. @bbondy can help with that

File: patches/chrome-installer-util-google_update_settings.cc.patch [hunk 3]
```diff
@@ -0,0 +1,61 @@
+diff --git a/chrome/installer/util/google_update_settings.cc b/chrome/installer/util/google_update_settings.cc
+index 42a2c0cc1b9576aa5f6c03e615e4b8450611831d..5142fdf31a4212b79a0d9d2ed99908b7c0e6820c 100644
+--- a/chrome/installer/util/google_update_settings.cc
++++ b/chrome/installer/util/google_update_settings.cc
+@@ -36,7 +36,11 @@ using base::win::RegKey;
+ using installer::InstallationState;
+ 
+ const wchar_t GoogleUpdateSettings::kPoliciesKey[] =
++#if defined(BRAVE_CHROMIUM_BUILD) && defined(OFFICIAL_BUILD)
```
Comment (bridiver) [inline patches/chrome-installer-util-google_update_settings.cc.patch]: same as below undefine/redefine override

File: patches/chrome-installer-mini_installer-appid.h.patch [hunk 4]
```diff
@@ -0,0 +1,13 @@
+diff --git a/chrome/installer/mini_installer/appid.h b/chrome/installer/mini_installer/appid.h
+index c408242814504aae086fc332f0d8bbd7872821a2..deadf14c9c8a17d7d83fcf3428cd3734a92d70b1 100644
+--- a/chrome/installer/mini_installer/appid.h
++++ b/chrome/installer/mini_installer/appid.h
```
Comment (bridiver) [inline patches/chrome-installer-mini_installer-appid.h.patch]: why are we modifying this when we already have brave_appid?

File: patches/chrome-installer-mini_installer-BUILD.gn.patch [hunk 5]
```diff
@@ -11,6 +11,15 @@ index c365088b90b7a91faf52f5b108f64621a219b11f..3be0b71dfb7c8694da3d30608a3be627
      "configuration.cc",
      "configuration.h",
      "decompress.cc",
+@@ -37,7 +37,7 @@ source_set("lib") {
+     "mini_installer.cc",
+     "mini_installer.h",
+     "mini_installer.rc",
+-    "mini_installer_constants.cc",
++    "//brave/installer/mini_installer/mini_installer_constants.cc",
```
Comment (bridiver) [inline patches/chrome-installer-mini_installer-BUILD.gn.patch]: if we are replacing the entire file it's better to use `chromium_src` which will work automatically

File: patches/chrome-installer-util-google_update_settings.cc.patch [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+     L"SOFTWARE\\Policies\\Google\\Update";
++#endif
+ const wchar_t GoogleUpdateSettings::kUpdatePolicyValue[] = L"UpdateDefault";
+ const wchar_t GoogleUpdateSettings::kDownloadPreferencePolicyValue[] =
+     L"DownloadPreference";
+@@ -50,7 +54,7 @@ const int GoogleUpdateSettings::kCheckPeriodOverrideMinutesMax =
+ 
+ const GoogleUpdateSettings::UpdatePolicy
+ GoogleUpdateSettings::kDefaultUpdatePolicy =
+-#if defined(GOOGLE_CHROME_BUILD)
```
Comment (bridiver) [inline patches/chrome-installer-util-google_update_settings.cc.patch]: are we doing this for all `GOOGLE_CHROME_BUILD` guards in this file? If so it would be easier to just define GOOGLE_CHROME_BUILD if BRAVE_CHROMIUM_BUILD && OFFICIAL_BUILD and then undef at the end

File: patches/chrome-installer-util-google_chrome_distribution.cc.patch [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ 
++// TODO(simonhong): Use brave's url.
+ base::string16 GetUninstallSurveyUrl() {
+   const wchar_t kSurveyUrl[] = L"https://support.google.com/chrome/"
+                                L"contact/chromeuninstall3?hl=$1";
+@@ -146,7 +147,11 @@ base::string16 GoogleChromeDistribution::GetAppDescription() {
+ }
+ 
+ std::string GoogleChromeDistribution::GetSafeBrowsingName() {
++#if defined(BRAVE_CHROMIUM_BUILD)
```
Comment (bridiver) [inline patches/chrome-installer-util-google_chrome_distribution.cc.patch]: why is this needed?

File: patches/chrome-installer-util-google_chrome_distribution.cc.patch [hunk 8]
```diff
@@ -0,0 +1,24 @@
+diff --git a/chrome/installer/util/google_chrome_distribution.cc b/chrome/installer/util/google_chrome_distribution.cc
+index cbbb0d5a7dc3dfdfb88a53c1ac3d50e7550ce177..c8079d730f391541d5395a40cb69ea9b31959236 100644
+--- a/chrome/installer/util/google_chrome_distribution.cc
++++ b/chrome/installer/util/google_chrome_distribution.cc
+@@ -47,6 +47,7 @@ base::string16 LocalizeUrl(const wchar_t* url) {
+   return base::ReplaceStringPlaceholders(url, language, NULL);
+ }
+ 
++// TODO(simonhong): Use brave's url.
```
Comment (bridiver) [inline patches/chrome-installer-util-google_chrome_distribution.cc.patch]: please don't patch for comments. Open a ticket instead

File: patches/chrome-installer-util-browser_distribution.cc.patch [hunk 9]
```diff
@@ -1,7 +1,16 @@
...
+index 33c105c968dc46b62ab39ee766f1ecce4449c358..7504f12b590cdaa063e5a31e55dda24adfc142c6 100644
 --- a/chrome/installer/util/browser_distribution.cc
 +++ b/chrome/installer/util/browser_distribution.cc
+@@ -54,7 +54,7 @@ BrowserDistribution* BrowserDistribution::GetOrCreateBrowserDistribution(
+ BrowserDistribution* BrowserDistribution::GetDistribution() {
+   BrowserDistribution* dist = NULL;
+ 
+-#if defined(GOOGLE_CHROME_BUILD)
++#if defined(GOOGLE_CHROME_BUILD) || (defined(BRAVE_CHROMIUM_BUILD) && defined(OFFICIAL_BUILD))
+   dist = GetOrCreateBrowserDistribution<GoogleChromeDistribution>(
```
Comment (bridiver) [inline patches/chrome-installer-util-browser_distribution.cc.patch]: can't we just patch here to create a BraveDistribution and get rid of most of the patches below?

File: patches/chrome-install_static-install_modes.cc.patch [hunk 10]
```diff
@@ -0,0 +1,34 @@
+diff --git a/chrome/install_static/install_modes.cc b/chrome/install_static/install_modes.cc
+index 42d9966254758305e0dc999b186aa7abe1ff762a..147dd51211c889bdcbde29ae0537a6e17d045dd8 100644
+--- a/chrome/install_static/install_modes.cc
++++ b/chrome/install_static/install_modes.cc
+@@ -12,6 +12,21 @@ std::wstring GetUnregisteredKeyPathForProduct(const wchar_t* product) {
+   return std::wstring(L"Software\\").append(product);
+ }
+ 
++#if defined(BRAVE_CHROMIUM_BUILD)
```
Comment (bridiver) [inline patches/chrome-install_static-install_modes.cc.patch]: override with undefine/redefine

File: patches/chrome-browser-ui-BUILD.gn.patch [hunk 11]
```diff
@@ -0,0 +1,17 @@
+diff --git a/chrome/browser/ui/BUILD.gn b/chrome/browser/ui/BUILD.gn
+index 72dde06a99a77b34f466cbfba3d478dc60b2d86c..293ded5d7615b4cc65fb739f7cde499ee39bf861 100644
+--- a/chrome/browser/ui/BUILD.gn
++++ b/chrome/browser/ui/BUILD.gn
+@@ -2669,6 +2669,12 @@ split_static_library("ui") {
+         "webui/settings/incompatible_applications_handler_win.h",
+       ]
+       deps += [ "//google_update" ]
++    } else if (brave_chromium_build && is_official_build) {
```
Comment (bridiver) [inline patches/chrome-browser-ui-BUILD.gn.patch]: can't we just add these in our own gn file?

File: win_build_output/midl/google_update/x64/google_update_idl.h [hunk 12]
```diff
@@ -0,0 +1,5747 @@
+
+
+/* this ALWAYS GENERATED file contains the definitions for the interfaces */
```
Comment (bridiver) [inline win_build_output/midl/google_update/x64/google_update_idl.h]: how is this file generated?

File: patches/chrome-browser-ui-BUILD.gn.patch (hunk 11 above)
Comment (bridiver) [reply patches/chrome-browser-ui-BUILD.gn.patch]: like you did here https://github.com/brave/brave-core/pull/109/files#diff-84f8700538180eec90bb2f4c9dff4da4R42

File: patches/chrome-installer-util-util_constants.cc.patch (hunk 1 above)
Comment (simonhong) [reply patches/chrome-installer-util-util_constants.cc.patch]: This var name comes from omaha.
To customize omaha for brave, company name should be changed to `BraveSoftware` because it is related with install directory. Then omaha makes env var by using it.
So, I think we should follow omaha var name in here.

File: patches/chrome-installer-mini_installer-appid.h.patch (hunk 4 above)
Comment (simonhong) [reply patches/chrome-installer-mini_installer-appid.h.patch]: We don't have copied header file.
Strangely, `brave_appid.cc` is not located in `chromium_src`. We want to move it under `chromium_src` right? also copy header file, too?

File: patches/chrome-installer-mini_installer-BUILD.gn.patch (hunk 5 above)
Comment (simonhong) [reply patches/chrome-installer-mini_installer-BUILD.gn.patch]: Ok, it would be good to move `//brave/installer/mini_installer/mini_installer_constants.cc` into `chromium_src`.

File: win_build_output/midl/google_update/x64/google_update_idl.h (hunk 12 above)
Comment (simonhong) [reply win_build_output/midl/google_update/x64/google_update_idl.h]: During the build, `midl.py` generates these file in temp directory and compares it with pre-installed version.

File: patches/chrome-browser-ui-BUILD.gn.patch [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
++    ]
++  }
+   allow_circular_includes_from +=
+       [ "//chrome/browser/ui/webui/bluetooth_internals" ]
+ 
+@@ -2661,6 +2666,7 @@ split_static_library("ui") {
+       "//ui/events:dom_keycode_converter",
+     ]
+ 
++    if (!brave_chromium_build) {
```
Comment (bridiver) [inline patches/chrome-browser-ui-BUILD.gn.patch]: isn't this redundant? brave_chrome_build will never be is_chrome_branded

Comment (simonhong) [reply patches/chrome-browser-ui-BUILD.gn.patch]: I used this because of `else` clause below.
This will disable `else` clause and all related things are handled in `brave/browser/ui/BUILD.gn` instead.

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 6 above)
Comment (bridiver) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: isn't the reverse just as likely?

File: patches/chrome-browser-ui-BUILD.gn.patch (hunk 13 above)
Comment (bridiver) [reply patches/chrome-browser-ui-BUILD.gn.patch]: but the else is just version_updater_basic, right? Do we really care if that gets built and not used? If it cases other problems then I think we should override the file to be empty in chromium_src

Comment (simonhong) [reply patches/chrome-browser-ui-BUILD.gn.patch]: I think basic implementation would be needed in non-official build, too.
In `AboutHandler`, it always reference it

Comment (bridiver) [reply patches/chrome-browser-ui-BUILD.gn.patch]: then can't we add it just like we did with google_update_win.cc ?

Comment (simonhong) [reply patches/chrome-browser-ui-BUILD.gn.patch]: `google_update_win` was only included in `google_chrome_build`.
So, I don't need to touch that.
But, `VersionUpdaterBasic` was included in non `google_chrome_build`.

File: win_build_output/midl/google_update/x64/google_update_idl.h (hunk 12 above)
Comment (bridiver) [reply win_build_output/midl/google_update/x64/google_update_idl.h]: then why are we copying the output? Shouldn't we be modifying the inputs so we get the correct output?

File: patches/components-update_client-updater_state_win.cc.patch (hunk 2 above)
Comment (bridiver) [reply patches/components-update_client-updater_state_win.cc.patch]: const overriding should work just like method overriding. same def/undef applies

Comment (bridiver) [reply patches/components-update_client-updater_state_win.cc.patch]: why would the result be the same as copying if we're only changing some of the values?

File: patches/chrome-browser-ui-BUILD.gn.patch (hunk 13 above)
Comment (bridiver) [reply patches/chrome-browser-ui-BUILD.gn.patch]: yep, I understand now, please just modify the patch as discussed

File: win_build_output/midl/google_update/x64/google_update_idl.h (hunk 12 above)
Comment (simonhong) [reply win_build_output/midl/google_update/x64/google_update_idl.h]: https://github.com/brave/brave-browser/pull/303 handles this.

File: patches/chrome-browser-ui-BUILD.gn.patch (hunk 13 above)
Comment (simonhong) [reply patches/chrome-browser-ui-BUILD.gn.patch]: done.

File: patches/components-update_client-updater_state_win.cc.patch (hunk 2 above)
Comment (simonhong) [reply patches/components-update_client-updater_state_win.cc.patch]: I referred `browser_about_handler.cc` in `chromium_src`.
In that file, upstream method name was changed and reused with additional implementation(`FixupBrowserAboutURLBraveImpl`).

I don't know how to adopt this technique to this file w/o changing functions implementation in this file.
Do you mean variable like re-define `kGoogleUpdatePoliciesKey` to `kGoogleUpdatePoliciesKey_ChromiumImpl`?

I think I don't know how to adopt undefine/redefine technique here.

Comment (bridiver) [reply patches/components-update_client-updater_state_win.cc.patch]: you really don't even have to redefine the old one here, I think you can just undef the old one and add a new one

Comment (simonhong) [reply patches/components-update_client-updater_state_win.cc.patch]: Do you mean like below?
```
#include "../../../../components/update_client/updater_state_win.cc"

#undef kGoogleUpdatePoliciesKey
#undef kCheckPeriodOverrideMinutes
#undef kUpdatePolicyValue
#undef kChromeUpdatePolicyOverride
#undef kRegPathGoogleUpdate 
#undef kRegPathClientsGoogleUpdate

namespace update_client {

namespace {

const wchar_t kGoogleUpdatePoliciesKey[] =
    L"SOFTWARE\\Policies\\BraveSoftware\\Update";
const wchar_t kCheckPeriodOverrideMinutes[] = L"AutoUpdateCheckPeriodMinutes";
const wchar_t kUpdatePolicyValue[] = L"UpdateDefault";
const wchar_t kChromeUpdatePolicyOverride[] =
    L"Update{AFE6A462-C574-4B8A-AF43-4CC60DF4563B}";
const wchar_t kRegPathGoogleUpdate[] = L"Software\\BraveSoftware\\Update";
const wchar_t kRegPathClientsGoogleUpdate[] =
    L"Software\\BraveSoftware\\Update\\Clients\\"
    L"{B131C935-9BE6-41DA-9599-1F776BEB8019}";

}  // namespace

}  // namespace update_client

```

Comment (simonhong) [reply patches/components-update_client-updater_state_win.cc.patch]: I did like this - https://github.com/brave/brave-core/pull/109/commits/b23256e3148924eb8ac3086aa97617fd3e4e47a8
Is this what you intended?

Comment (bridiver) [reply patches/components-update_client-updater_state_win.cc.patch]: does that work correctly? It doesn't seem like it would

Comment (bridiver) [reply patches/components-update_client-updater_state_win.cc.patch]: I didn't notice that they were inside an anonymous namespace and that might be problematic for undef. If that builds without error then I don't think it's using the new values

Comment (bridiver) [reply patches/components-update_client-updater_state_win.cc.patch]: but whether we patch or do something else, we only need the values we are overriding. We should not include things like kCheckPeriodOverrideMinutes or kCheckPeriodOverrideMinutesMax that we are not changing

Comment (simonhong) [reply patches/components-update_client-updater_state_win.cc.patch]: > does that work correctly? It doesn't seem like it would

It doesn't work

Comment (simonhong) [reply patches/components-update_client-updater_state_win.cc.patch]: Do you want to apply override technique like this (https://github.com/brave/brave-core/commit/b23256e3148924eb8ac3086aa97617fd3e4e47a8) to other files in this PR?

Comment (bridiver) [reply patches/components-update_client-updater_state_win.cc.patch]: well, it won't work if they're in an anonymous namespace like that. There's not way to reference/override them afaik

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 6 above)
Comment (simonhong) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: done with `#define GOOGLE_CHROME_BUILD`

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 3 above)
Comment (simonhong) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: Can you advice how to override variable? I still don't have any idea how to do it with separate file in `chormium_src`.

File: patches/chrome-install_static-install_modes.cc.patch (hunk 10 above)
Comment (simonhong) [reply patches/chrome-install_static-install_modes.cc.patch]: can't do for functions that are used by other functions in the same file.

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 3 above)
Comment (simonhong) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: `kPoliciesKey` is used by other methods in this file.
If I re-define `kPoliciesKey`, I think all methods that uses `kPoliciesKey` should be re-implemented.

Comment (simonhong) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: I think it's difficult to use override technique to member var/method.

I tried to override `ClassA::MethodA`, in new `class_a.cc` in `chromium_src` like below.
```
#define MethodA MethodA_ChromiumImpl
#include ".../class_a.cc"
#undef MethodA

void ClassA::MethodA() {
...
}
```
Compiler will complain like this `error: no member named MethodA in ClassA`.
I think the reason is header file(class_a.h) included in this file was also affected by macro.
So, there is no `MethodA` in this file.
Overriding member variable will have same error.
@bbondy

**Comment by simonhong** [comment]
I think define/undef/redefine technique is only useful when unmodified upstream method is re-used by redefined method.
In this pr, upstream method/variable is changed.
IMO, I can't apply this technique to this PR.
PTAL again.

File: patches/chrome-install_static-install_modes.cc.patch (hunk 10 above)
Comment (bbondy) [reply patches/chrome-install_static-install_modes.cc.patch]: you can do it for functions that are used by other functions in the same file, but you would also have to do the functions that use that function. You can just only do the functions that use that function I think in this case.

File: patches/chrome-installer-util-google_update_settings.cc.patch [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+ GoogleUpdateSettings::kDefaultUpdatePolicy =
+ #if defined(GOOGLE_CHROME_BUILD)
+@@ -913,3 +922,7 @@ bool GoogleUpdateSettings::ReadExperimentLabels(
+ 
+   return result == ERROR_SUCCESS;
+ }
++
++#if defined(BRAVE_CHROMIUM_BUILD) && defined(OFFICIAL_BUILD)
++#undef GOOGLE_CHROME_BUILD
++#endif
```
Comment (bbondy) [inline patches/chrome-installer-util-google_update_settings.cc.patch]: I don't think you need to worry about this since you're only defining it in the single .cc file.

File: patches/chrome-installer-util-google_update_settings.cc.patch [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+ const wchar_t GoogleUpdateSettings::kDownloadPreferencePolicyValue[] =
+     L"DownloadPreference";
+@@ -48,6 +52,11 @@ const wchar_t GoogleUpdateSettings::kCheckPeriodOverrideMinutes[] =
+ const int GoogleUpdateSettings::kCheckPeriodOverrideMinutesMax =
+     60 * 24 * 7 * 6;
+ 
++
++#if defined(BRAVE_CHROMIUM_BUILD) && defined(OFFICIAL_BUILD)
++#define GOOGLE_CHROME_BUILD
++#endif
```
Comment (bbondy) [inline patches/chrome-installer-util-google_update_settings.cc.patch]: Override method can just define this at the top I think.

File: patches/chrome-installer-util-google_update_settings.cc.patch [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
++++ b/chrome/installer/util/google_update_settings.cc
+@@ -36,7 +36,11 @@ using base::win::RegKey;
+ using installer::InstallationState;
+ 
+ const wchar_t GoogleUpdateSettings::kPoliciesKey[] =
++#if defined(BRAVE_CHROMIUM_BUILD) && defined(OFFICIAL_BUILD)
++    L"SOFTWARE\\Policies\\BraveSoftware\\Update";
++#else
+     L"SOFTWARE\\Policies\\Google\\Update";
++#endif
```
Comment (bbondy) [inline patches/chrome-installer-util-google_update_settings.cc.patch]: Continue to patch for this part, later I think we will add some new way to do string replacements for overrides.  I imagine it something like we copy from chromium_src to some gen dir, and we include from there after applying some overrides.   Different task though.

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 3 above)
Comment (bbondy) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: replied below.

File: patches/chrome-installer-util-util_constants.cc.patch (hunk 1 above)
Comment (bbondy) [reply patches/chrome-installer-util-util_constants.cc.patch]: can be done with the later string substitution method.

File: patches/chrome-installer-util-browser_distribution.cc.patch (hunk 9 above)
Comment (bbondy) [reply patches/chrome-installer-util-browser_distribution.cc.patch]: ok

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 15 above)
Comment (simonhong) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: Got it, placing this on top of overridden file will be better.

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 14 above)
Comment (simonhong) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: This would be needed when we enable jumbo build later.

Comment (bbondy) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: Good point, I'll watch out for that in the future on reviews too.

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 15 above)
Comment (simonhong) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: If we add this def to overridden file, all header files included in `google_update_settings.cc`  could be also affected.
It can cause mismatch of header member in some files.

File: patches/chrome-install_static-install_modes.cc.patch (hunk 10 above)
Comment (simonhong) [reply patches/chrome-install_static-install_modes.cc.patch]: Left some comment for the record.
This is the reason why patching technique is used instead of overriding.
In case of `install_modes.cc`, override technique causes modifying all functions in that file and can't reuse upstream functions. It is similar to copying technique.
IMO, patching technique is more better to copying for future rebase.
I thought def/undef technique is *only* useful when we can re-use upstream implementation

File: patches/chrome-installer-util-google_update_settings.cc.patch (hunk 15 above)
Comment (bridiver) [reply patches/chrome-installer-util-google_update_settings.cc.patch]: actually don't we also want it enabled there? It looks like we do to me.
